### PR TITLE
Permission module with decorator

### PR DESCRIPTION
**Pull request by cobusc**
This PR introduces a `@require_permissions` decorator that we will use to decorate API calls to ensure that only users with the appropriate permissions can execute certain actions. There is also a comprehensive test suite.

In order to complete the task I had to put down some solutions for caching definitions from the Access Control component, etc. We will need to revisit these implementations at a later stage to ensure it is the best implementation. 

The focus of this ticket was the permission module, so please focus most of your attention on the following files when PRing:
* `management_layer/permission/__init__.py`
* `management_layer/permission/utils.py`
* `management_layer/permission/test_utils.py`

### CODE REVIEW

File: management_layer/mappings.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
+"""
+This module manages the common permission-related information in convenient
+formats.
+
+In order to prevent repeated API lookups, data can be loaded when the
+application is started and refreshed as necessary. We typically store
+dictionaries where the key is the id of the entity and the value the entity
```
Comment (AltusBarry) [inline management_layer/mappings.py]: `value the` -> `value of the`

File: management_layer/permission/__init__.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        @wraps(f)
+        def wrapped_f(*args, **kwargs):
+            """
+            The purpose of the wrapper is to get the
+            :param args: A list of positional arguments
+            :param kwargs: A dictionary of keyword arguments
+            :return: Whatever the wrapped function
+            """
+            # Extract the user UUID from the function arguments
+            user_field_type = type(user_field)
```
Comment (ruands) [inline management_layer/permission/__init__.py]: Couldn't you just use `if type(user_field) is int`? It does not look like `user_field_type` is being used somewhere else after this.

File: management_layer/permission/__init__.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                user = kwargs[user_field]
+            else:
+                raise RuntimeError("Invalid value for user_field")
+
+            # Validate the UUID
+            if not isinstance(user, UUID):
+                raise RuntimeError("User is expected to be a UUID")
+
+            # Extract the site UUID from the function arguments
+            site_field_type = type(site_field)
```
Comment (ruands) [inline management_layer/permission/__init__.py]: Same as the above.

File: management_layer/mappings.py (hunk 1 above)
Comment (cobusc) [reply management_layer/mappings.py]: I think the sentence is correct, although I can maybe rephrase it better:
* key => id of entity
* value => the entity itself
How about "...where the key is the id of the entity, and the value is the entity itself."?

File: management_layer/permission/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,683 @@
...
+        with self.assertRaises(utils.Forbidden):
+            empty_any(valid_uuid, valid_uuid)
+
+    @patch("management_layer.permission.utils.get_user_roles_for_site")
+    def test_positional_args(self, mocked_function):
+        """
+        Check that positional argument lookups of the user and site
+        information works.
+        We mock a response for the get_user_roles_for_site() function in
+        order check that it was called with the appropriate values.
```
Comment (ruands) [inline management_layer/permission/test_utils.py]: `in order to check`

File: management_layer/permission/utils.py [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+    :param user: The UUID of the user performing the action
+    :param operator: any or all
+    :param resource_permissions: The resource permissions required
+    :param site: The site from which the user is making the request
+    :param domain: The domain for which the user is making the request
+    :param nocache: Bypass the cache if True
+    :return: True if the user is has the required permissions on the resources
+    """
+    # Either a site or domain needs to be provided
+    if bool(site) == bool(domain):
```
Comment (AltusBarry) [inline management_layer/permission/utils.py]: Shouldn't this then read `if bool(site) or bool(domain)`?

File: management_layer/permission/__init__.py (hunk 2 above)
Comment (cobusc) [reply management_layer/permission/__init__.py]: `user_field_type` is used twice: in the `if` on line 91 and the `elif` on line 93. I consider it best practice to avoid unnecessary function calls, hence this implementation.

File: management_layer/permission/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,683 @@
...
+        # ...and verify that the mocked function was called with the right
+        # arguments.
+        mocked_function.assert_called_with(user, site, False)
+
+    @patch("management_layer.permission.utils.get_user_roles_for_site")
+    def test_keyword_args(self, mocked_function):
+        """
+        Check that keyword-based lookups of the user and site information works.
+        We mock a response for the get_user_roles_for_site() function in
+        order check that it was called with the appropriate values.
```
Comment (ruands) [inline management_layer/permission/test_utils.py]: `in order to check`

File: management_layer/permission/__init__.py (hunk 3 above)
Comment (cobusc) [reply management_layer/permission/__init__.py]: Same as above

File: management_layer/permission/test_utils.py [hunk 7]
```diff
@@ -0,0 +1,683 @@
...
+        # arguments.
+        mocked_function.assert_called_with(user, site, False)
+
+    @patch("management_layer.permission.utils.get_user_roles_for_site")
+    def test_mixed_args_nocache(self, mocked_function):
+        """
+        Check that a mix of positional and keyword-based lookups of user and
+        site information works.
+        We mock a response for the get_user_roles_for_site() function in
+        order check that it was called with the appropriate values.
```
Comment (ruands) [inline management_layer/permission/test_utils.py]: Copy and paste of this docstring carried this error far :)

File: management_layer/permission/utils.py [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+    Bypassing the cache can have a significant performance and should only be
+    used in exceptional circumstances.
+    :param user: The user
+    :param site: The site
+    :param domain: The domain
+    :param nocache: Bypass the cache if True
+    :return: A list of roles ids
+    """
+    # Either a site or domain needs to be provided
+    if bool(site) == bool(domain):
```
Comment (AltusBarry) [inline management_layer/permission/utils.py]: See previous comment, assuming I am not misunderstanding the check.

File: management_layer/permission/utils.py (hunk 5 above)
Comment (cobusc) [reply management_layer/permission/utils.py]: No. Using `or` would imply a logical OR operation. What I am doing here is a XOR (eXlusive-OR).
* If neither `site` nor `domain` are specified, we have `False == False` which will raise the error.
* If both `site` and `domain` are specified, we have `True == True` which will raise the error.
In the other cases we have `True == False` and `False == True`, neither of which will raise an error.

I will change the comment to be more explicit: `Either a site or domain needs to be provided, but not both.`

File: management_layer/permission/__init__.py (hunk 2 above)
Comment (ruands) [reply management_layer/permission/__init__.py]: I thought it may be because of having to call the function twice.

File: management_layer/permission/utils.py (hunk 8 above)
Comment (cobusc) [reply management_layer/permission/utils.py]: Refer to my reply for the previous comment.

File: management_layer/permission/test_utils.py (hunk 4 above)
Comment (cobusc) [reply management_layer/permission/test_utils.py]: :+1:

File: management_layer/permission/utils.py (hunk 8 above)
Comment (cobusc) [reply management_layer/permission/utils.py]: Yeah, check my previous answer.

File: management_layer/mappings.py (hunk 1 above)
Comment (AltusBarry) [reply management_layer/mappings.py]: Yes please.

File: management_layer/permission/utils.py (hunk 5 above)
Comment (AltusBarry) [reply management_layer/permission/utils.py]: Thanks, makes sense.

**Comment by AltusBarry** [comment]
+1 after comment updates. Should the overall travis builds be passing yet?
